From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Damglador <vse.stopchanskyi@gmail.com>
Date: Mon, 27 Oct 2025 23:35:43 +0100
Subject: fix: allow setting tray id

This is needed to allow setting custom tray id when initializing a tray icon.
With current behaviour all programs get chrome_status_icon_1 as their id in tray.
So tray can't tell apart Electron apps,
which introduces issues like https://bugs.kde.org/show_bug.cgi?id=470840.

This patch can be removed after being upstreamed. See discussion at
https://github.com/electron/electron/pull/48675#issuecomment-3452781711
for more info.

diff --git a/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.cc b/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.cc
index 60a3c13471a630c36b377544e83bd5bf967c47d7..8169dfff20f281ebe9b489fb0449fbd09af7d0b8 100644
--- a/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.cc
+++ b/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.cc
@@ -137,8 +137,9 @@ int NextServiceId() {
   return ++status_icon_count;
 }
 
-std::string PropertyIdFromId(int service_id) {
-  return "chrome_status_icon_" + base::NumberToString(service_id);
+std::string PropertyIdFromId(int service_id, const std::string_view app_name) {
+  return base::StrCat(
+      {app_name, "_status_icon_", base::NumberToString(service_id)});
 }
 
 using DbusImage = std::tuple</*width=*/int32_t,
@@ -249,15 +250,17 @@ scoped_refptr<dbus::Bus> GetBusForNewStatusIcon() {
 
 }  // namespace
 
-StatusIconLinuxDbus::StatusIconLinuxDbus()
-    : StatusIconLinuxDbus(GetBusForNewStatusIcon()) {}
+StatusIconLinuxDbus::StatusIconLinuxDbus(const std::string_view app_name)
+    : StatusIconLinuxDbus(GetBusForNewStatusIcon(), app_name) {}
 
-StatusIconLinuxDbus::StatusIconLinuxDbus(scoped_refptr<dbus::Bus> bus)
+StatusIconLinuxDbus::StatusIconLinuxDbus(scoped_refptr<dbus::Bus> bus,
+                                         const std::string_view app_name)
     : bus_(std::move(bus)),
       should_write_icon_to_file_(ShouldWriteIconToFile()),
       icon_task_runner_(base::ThreadPool::CreateSequencedTaskRunner(
           {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
-           base::TaskShutdownBehavior::BLOCK_SHUTDOWN})) {
+           base::TaskShutdownBehavior::BLOCK_SHUTDOWN})),
+      app_name_(app_name) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
   if (bus_) {
     CheckStatusNotifierWatcherHasOwner();
@@ -469,7 +472,8 @@ void StatusIconLinuxDbus::OnHostRegisteredResponse(
   SetProperty<"s">(kPropertyAttentionIconName, "", false);
   SetProperty<"s">(kPropertyAttentionMovieName, "", false);
   SetProperty<"s">(kPropertyCategory, kPropertyValueCategory, false);
-  SetProperty<"s">(kPropertyId, PropertyIdFromId(service_id_), false);
+  SetProperty<"s">(kPropertyId, PropertyIdFromId(service_id_, app_name_),
+                   false);
   SetProperty<"s">(kPropertyOverlayIconName, "", false);
   SetProperty<"s">(kPropertyStatus, kPropertyValueStatus, false);
   SetProperty<"s">(kPropertyTitle, "", false);
diff --git a/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.h b/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.h
index 5f7e20bc0b56b5621cfb04d96712b2f1db91cf0e..9668ee5a48e180d0ace989c6cb6d9364a162c5ce 100644
--- a/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.h
+++ b/chrome/browser/ui/views/status_icons/status_icon_linux_dbus.h
@@ -38,8 +38,9 @@ class StatusIconLinuxDbus : public ui::StatusIconLinux,
                             public ui::SimpleMenuModel::Delegate,
                             public base::RefCounted<StatusIconLinuxDbus> {
  public:
-  StatusIconLinuxDbus();
-  explicit StatusIconLinuxDbus(scoped_refptr<dbus::Bus> bus);
+  StatusIconLinuxDbus(const std::string_view app_name = "chrome");
+  explicit StatusIconLinuxDbus(scoped_refptr<dbus::Bus> bus,
+                               const std::string_view app_name = "chrome");
 
   StatusIconLinuxDbus(const StatusIconLinuxDbus&) = delete;
   StatusIconLinuxDbus& operator=(const StatusIconLinuxDbus&) = delete;
@@ -161,6 +162,8 @@ class StatusIconLinuxDbus : public ui::StatusIconLinux,
   size_t icon_file_id_ = 0;
   base::FilePath icon_file_;
 
+  std::string app_name_;
+
   base::WeakPtrFactory<StatusIconLinuxDbus> weak_factory_{this};
 };
 
